src/kernel/boot.c

```diff
@@ -873,21 +881,60 @@ BOOT_CODE bool_t init_freemem(word_t n_available, const p_region_t *available,
      * from. If possible, retain any left over memory as an extra p region */
     word_t size = calculate_rootserver_size(it_v_reg, extra_bi_size_bits);
     word_t max = rootserver_max_size_bits(extra_bi_size_bits);
+    pptr_t start;
     for (; i >= 0; i--) {
-        word_t next = i + 1;
```

Comment:
This was probably here to avoid SimplExportAndRefine's flaky handling of array index expressions. I would try reinstating this `next` variable to see if that fixes the SimplExportAndRefine failure.

If that works, could you add a comment, e.g.:
```
// This avoids SimplExportAndRefine's poor handling of array indices that
// are sums of variables and constants.
```

My apologies if it was me that previously added `next` without such a comment!

Reply:
> This was probably here to avoid SimplExportAndRefine's flaky handling of array index expressions. I would try reinstating this `next` variable to see if that fixes the SimplExportAndRefine failure.

I just went ahead and pushed a commit that does this. We'll see how it goes...

Reply:
Ugh, sorry for all the noise. That's what I get for trying to make a small change while in a meeting, and using GitHub as my compiler.

Reply:
That didn't work, so I don't know why SimplExportAndRefine is failing. I'm not sure it's worth looking into it more deeply, so I suggest trying to minimise the code changes in this PR.

Some general comments, now that I've looked more closely at the PR...

Firstly, thanks for finding the issue, and proposing a fix!

Comments are worth adding to this code, but I think they could be more concise. May I recommend explicitly stating (and where reasonable, asserting) loop invariants, e.g.:
```
/* Invariant: both i and (i + 1) are valid indices in ndks_boot.freemem. */
assert(i < ARRAY_SIZE(ndks_boot.freemem) - 1);
/* Invariant; the region at index i is the current candidate.
 * Invariant: regions 0 up to (i - 1), if any, are additional candidates.
 * Invariant: region (i + 1) is undefined, and should not be read.
 * Invariant: regions above (i + 1), if any, are empty or too small to use.
 * Invariant: defined regions (all except i + 1) are ordered,
 *            non-overlapping and unallocated. */
```
Having written down these invariants, it's immediately clear that two of the conditionals in the loop (prior to this PR) are redundant, and it's easier to explain some other things. I'll post what I mean separately...

However, I'm not sure the other changes in this PR actually make things clearer. When all else is equal, concision is preferable to verbosity, and I'm not convinced this PR has a payoff to compensate for the increased verbosity. On the contrary, the PR (slightly) increases the cognitive distance between things that readers need to know are connected. For example, when looking at the code that shuffles the current region up to the unused slot, we now need to remember which slots `reg` and `next` point to. IMO, the array expressions they replace are not so cumbersome that they warrant fresh names. Also, I think the code to perform the root server setup is not so big that it deserves to be placed after the loop, where the reader needs to figure out how to differentiate between success and failure cases.

As a point of comparison, here's an alternative version that I think clarifies how the code works with minimal changes to it: https://github.com/seL4/seL4/pull/587

I'm not sure if the `i + 1` will make it through SimplExportAndRefine, but worth a try...